nx/source/services/ro.c

```diff
@@ -0,0 +1,284 @@
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 0;
+    raw->pid = 0;
+    raw->nro_address = nro_address;
+    raw->nro_size = nro_size;
+    raw->bss_address = bss_address;
+    raw->bss_size = bss_size;
+
+    Result rc = serviceIpcDispatch(&g_roSrv);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+            u64 out_address;
+        } *resp;
+
+        serviceIpcParse(&g_roSrv, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+
+        if (R_SUCCEEDED(rc)) {
+            if (out_address) *out_address = resp->out_address;
+        }
+    }
+
+    return rc;
+}
+
+Result ldrRoUnloadNro(u64 nro_address) {
+    IpcCommand c;
+    ipcInitialize(&c);
+    ipcSendPid(&c);
+
+    struct {
+        u64 magic;
+        u64 cmd_id;
+        u64 pid;
+        u64 nro_address;
+    } *raw;
+
+    raw = serviceIpcPrepareHeader(&g_roSrv, &c, sizeof(*raw));
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 1;
+    raw->pid = 0;
+    raw->nro_address = nro_address;
+
+    Result rc = serviceIpcDispatch(&g_roSrv);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+        } *resp;
+
+        serviceIpcParse(&g_roSrv, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+    }
+
+    return rc;
+}
+
+Result ldrRoLoadNrr(u64 nrr_address, u64 nrr_size) {
+    IpcCommand c;
+    ipcInitialize(&c);
+    ipcSendPid(&c);
+
+    struct {
+        u64 magic;
+        u64 cmd_id;
+        u64 pid;
+        u64 nrr_address;
+        u64 nrr_size;
+    } *raw;
+
+    raw = serviceIpcPrepareHeader(&g_roSrv, &c, sizeof(*raw));
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 2;
+    raw->pid = 0;
+    raw->nrr_address = nrr_address;
+    raw->nrr_size = nrr_size;
+
+    Result rc = serviceIpcDispatch(&g_roSrv);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+        } *resp;
+
+        serviceIpcParse(&g_roSrv, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+    }
+
+    return rc;
+}
+
+Result ldrRoUnloadNrr(u64 nrr_address) {
+    IpcCommand c;
+    ipcInitialize(&c);
+    ipcSendPid(&c);
+
+    struct {
+        u64 magic;
+        u64 cmd_id;
+        u64 pid;
+        u64 nrr_address;
+    } *raw;
+
+    raw = serviceIpcPrepareHeader(&g_roSrv, &c, sizeof(*raw));
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 3;
+    raw->pid = 0;
+    raw->nrr_address = nrr_address;
+
+    Result rc = serviceIpcDispatch(&g_roSrv);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+        } *resp;
+
+        serviceIpcParse(&g_roSrv, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+    }
+
+    return rc;
+}
+
+Result _ldrRoInitialize(void) {
```

Comment:
static

Reply:
static only needs to be placed in the prototype, not in the function definition. In fact you get a warning iirc.